scripts/casefold_db.py

```diff
@@ -0,0 +1,285 @@
+        db.commit()
+
+
+def update_global_assoc(
+    sydent, db: sqlite3.Connection, send_email: bool, dry_run: bool
+):
+    """Update the DB table global_threepid_associations so that all stored
+    emails are casefolded, the signed association is re-signed and any duplicate
+    mxid's associated with the given email are deleted.
+
+    :return: None
+    """
+
+    # get every row where the local server is origin server and medium is email
+    origin_server = sydent.server_name
+    medium = "email"
+
+    cur = db.cursor()
+    res = cur.execute(
+        "SELECT address, mxid, sgAssoc FROM global_threepid_associations WHERE medium = ?"
+        "AND originServer = ? ORDER BY ts DESC",
+        (medium, origin_server),
+    )
+
+    # dict that stores email address with mxid, email address, lookup hash, and
+    # signed association
+    associations: Dict[str, List[Tuple[str, str, str, str]]] = {}
+
+    # iterate through selected associations, casefold email, rehash it, re-sign the
+    # associations and add to associations dict
+    for address, mxid, sg_assoc in res.fetchall():
+        casefold_address = address.casefold()
+
+        # rehash the email since hash functions are case-sensitive
+        lookup_hash = calculate_lookup_hash(sydent, casefold_address)
+
+        # update signed associations with new casefolded address and re-sign
+        sg_assoc = json_decoder.decode(sg_assoc)
+        sg_assoc["address"] = address.casefold()
+        sg_assoc = json.dumps(
+            signedjson.sign.sign_json(
+                sg_assoc, sydent.server_name, sydent.keyring.ed25519
+            )
+        )
+
+        if casefold_address in associations:
+            associations[casefold_address].append(
+                (address, mxid, lookup_hash, sg_assoc)
+            )
+        else:
+            associations[casefold_address] = [(address, mxid, lookup_hash, sg_assoc)]
+
+    # list of arguments to update db with
+    db_update_args: List[Tuple[Any, str, str, str, str]] = []
+
+    # list of mxids to delete
+    to_delete: List[Tuple[str]] = []
+
+    # list of mxids and addresses to send emails to letting them know the mxid
+    # has been deleted
+    mxids: List[Tuple[Any, Any]] = []
+
+    for casefold_address, assoc_tuples in associations.items():
+        db_update_args.append(
+            (
+                casefold_address,
+                assoc_tuples[0][2],
+                assoc_tuples[0][3],
+                assoc_tuples[0][0],
+                assoc_tuples[0][1],
+            )
+        )
+
+        if len(assoc_tuples) > 1:
+            # Iterate over all associations except for the first one, since we've already
+            # processed it.
+            for address, mxid, _, _ in assoc_tuples[1:]:
+                to_delete.append((address,))
+                mxids.append((mxid, address))
+
+    # iterate through the mxids and send email, let's only send on email per mxid
+    if send_email and not dry_run:
```

Comment:
I don't think we want to send emails when updating the global table, because associations in this table either come from the local server (which we're already taking care of) or any server replicating to it (which should be under the supervision of the same admin and should be updated around the same time), and it also likely means that every user would receive two emails, one for the local table, one for the global one.